certbot-ci/certbot_integration_tests/certbot_tests/assertions.py

```diff
@@ -0,0 +1,81 @@
+"""This module contains advanced assertions for the certbot integration tests."""
+import os
+import grp
+
+
+# Defining __all__ allows to make a static import "from assertions import *" without importing
+# also the modules that are part of this module.
+__all__ = ['assert_hook_execution', 'assert_save_renew_hook', 'assert_certs_count_for_lineage',
+           'assert_equals_permissions', 'assert_equals_group_owner', 'assert_world_permissions']
+
+
+def assert_hook_execution(probe_path, probe_content):
+    """
+    Assert that a certbot hook has been executed
+    :param probe_path: path to the file that received the hook output
+    :param probe_content: content expected when the hook is executed
+    """
+    with open(probe_path, 'r') as file:
+        lines = file.readlines()
+
+    # Comparing pattern to each line avoids to match "pre" for a line with "pre-override"
+    assert [line for line in lines if line == '{0}{1}'.format(probe_content, os.linesep)]
+
+
+def assert_save_renew_hook(config_dir, lineage):
+    """
+    Assert that the renew hook configuration of a lineage has been saved.
+    :param config_dir: location of the certbot configuration
+    :param lineage: lineage domain name
+    """
+    assert os.path.isfile(os.path.join(config_dir, 'renewal/{0}.conf'.format(lineage)))
+
+
+def assert_certs_count_for_lineage(config_dir, lineage, count):
+    """
+    Assert the number of certificates generated for a lineage.
+    :param config_dir: location of the certbot configuration
+    :param lineage: lineage domain name
+    :param count: number of expected certificates
+    """
+    archive_dir = os.path.join(config_dir, 'archive')
+    lineage_dir = os.path.join(archive_dir, lineage)
+    certs = [file for file in os.listdir(lineage_dir) if file.startswith('cert')]
+    assert len(certs) == count
+
+
+def assert_equals_permissions(file1, file2, mask):
+    """
+    Assert that permissions on two files are identical in respect to a given umask.
+    :param file1: first file path to compare
+    :param file2: second file path to compare
+    :param mask: umask to apply before comparing file modes
+    """
+    mode_file1 = os.stat(file1).st_mode & mask
+    mode_file2 = os.stat(file2).st_mode & mask
+
+    assert mode_file1 == mode_file2
+
+
+def assert_equals_group_owner(file1, file2):
+    """
+    Assert that two files have the same group owner.
+    :param file1: first file path to compare
+    :param file2: second file path to compare
+    :return:
+    """
+    group_owner_file1 = grp.getgrgid(os.stat(file1).st_gid)[0]
```

Comment:
`grp.getgrgid(id)[0]` gives the group name corresponding to the group whose the group id is `id`.

Reply:
Can you explain more about what's happening in Windows here?

Reply:
I see the underlying point here. This code is specific to Linux, so it can stay to a comparison between gids. The full implementation here will be to have a branch code to execute either the Linux specific code, or the Windows specific code.

To be more precise for groups, on Windows the assertion will never fail because there is no notion of Posix group on it. 

I will change the code tomorrow.

Note that I will not do the Windows part here. Indeed, my approach is to write cross platform compatible code if possible, but write Windows specific code in a future PR, when the full code of current spec of Certbot-CI will be merged.

Reply:
Gotcha, thanks. Are there other places in here where something weird is being done that's not strictly necessary now, but makes sense because of future planned windows compatibility? I'd love to see those marked.

Reply:
Not that I could remember of. After checking the code, it seems not also. I wrote specifically the shell scripts in Python to be cross platform, but beside that, I did not take any specific effort. This here was an exception, I remembered been influenced at the time of writing this assertion, that arrived quite late from another PR made by Sydney, by the ongoing work on Windows file permissions. 

But it is not justified to think about Windows at this time. It will when every functional code in certbot core for windows has landed, and Certbot-CI is fully functional on Linux.